File: src/com/example/algorithm/sorting/QuickSort.java
package com.example.algorithm.sorting;

import java.util.Arrays;

public class QuickSort {


    public static int iteration = 0;

    public static void main(String[] args) {

        iteration=0;
        int[] data = { 8, 7, 2, 1, 0, 9, 6 , 5, 4, 11};
        System.out.println(Arrays.toString(data));
        quickSort(data, 0, data.length-1);
        System.out.println(Arrays.toString(data));
        System.out.println("Quick Sort for "+data.length+" size data took total of iteration "+iteration);

    }


    static void quickSort(int[] arr, int low, int high){
        if (low<high) {
            int m = pivote(arr, low, high);
            quickSort(arr, low, m-1);
            quickSort(arr,  m+1, high);
        }
    }

     static int pivote(int[] arr, int low, int high) {

        int piv = arr[high];
        int i = low-1;
        for (int j=low; j<=high; j++){
            iteration++;
            if (arr[j]<=piv){
                i++;

                int temp= arr[j];
                arr[j] = arr[i];
                arr[i]= temp;
            }
        }
        return i;
    }
}

File: src/com/example/algorithm/QuickSort.kt
package com.example.algorithm

class QuickSort {


    companion object{
        @JvmStatic
        fun main(args: Array<String>){
            val arr = intArrayOf(3, 2, 6, 5, 9, 4, 1)
            arr.forEach { print("$it, ") }
            println()
            QuickSort().quickSort(arr, 0, arr.size-1)
            arr.forEach { print("$it, ") }
        }

    }

    fun quickSort(array: IntArray, l: Int, h: Int) {

        if (l<h) {
            val pivot = getPivot(array, l, h)
            quickSort(array, l, pivot-1)
            quickSort(array, pivot + 1, h)
        }
    }

    private fun getPivot(array: IntArray, l: Int, h: Int): Int {
        val p = array[h]
        var i = l - 1
        for (j in l..h) {
            if (array[j] <= p) {
                i++
                val temp = array[j]
                array[j] = array[i]
                array[i] = temp
            }
        }
        return i
    }
}